### ISSUE TITLE
Bloat in H5 storage following repeated SIGKILL

### ISSUE THREAD

**Issue by root-11** [root-11/tablite#40]
# Important notes concerning multiprocessing, Startup and Exit.

For multiprocessing to be efficient, subprocesses must be able to create tables
in the same storage as main, as this avoids the need for moving the data.

Zero entropy = Zero copy.
------------------------

As `Table`'s refer `Column`'s which refer to `Page`'s, which in turn contain
data, each process must keep track of how each columns point to each page

This takes premise in a weakvaluedictionary called the `registry`, where each
`Page` has a reference count.

When a table is created and columns are added, pages may be empty. However if a
page is "copied" only the metadata (the pointer) is copied.

When the "copy" is deleted, only the metadata is deleted, as the original table
still has a reference to the data. For the reference system to remain valid,
the reference count in the `registry` must obviously be decremented.

Multi processing
----------------

When a subprocess exits (SIGTERM), python performs garbage collection (GC).
This means that any variable pointing to a table created by the subprocess
will also be deleted. This triggers a call to `table.__del__(self)`.

As the main process might be busy receiving data from other subprocesses, the
`__del__` call can happen before main has taken ownership of the data.

To assure that the output from the subprocess isn't lost, the subprocess marks
the table as `table.saved=True` far in advance of garbage collection. The logic
is simple:

    def __del__(self):
        decrement reference count  # ignore if unknown.
        if saved != True:
            delete all data.

This guarantees that the call to `__del__` doesn't erase the data from storage,
and thereby permits `__main__` to collect it after the subprocess has exited.

When `main` loads the data using the subprocess' table key, the reference count
in `main` will be correct as `Table.load(path, key)` increments the reference 
count.

If `main` sets `saved = False` on the table, that the subprocess created before
it exited, it will now be deleted.

Should there be a error in the implementation of the interaction with the sub-
process, so that `main` does not set `saved = False` after loading, then the
table will never be deleted.

To help with the clean up there are two options:

1. `Table.reset_storage()`, which truncates the storage to 0 bytes. E.g. 
100% dataloss. This is a pragmatic solution during testing where pre-existing
state must be avoided.

2. `Table.load_stored_tables()`, which returns a list of all tables in the 
storage (discussed in 5, below). 
A user can now loop through all tables and set
`saved=False` and exit (SIGTERM) which will garbage collect all loaded tables
as `Table.load_stored_tables()` increments the reference counter from zero to
one for all tables, changes `save` and then drops them.

This is the preferred method whenever something unrelated to `tablite` caused
the python process exit.

Whilst these options are convenient, there are a couple of edge cases, that 
should be considered first:

    3. If another terminal is opened which uses the same H5 file.
    4. If a terminal is killed (SIGKILL)
    5. Two terminals use the same import.

3. As there is no conflict with another terminal using the same store, calling
`Table.reset_storage()` will have an inconvenient side effect, as it erases all 
data - including any created by the other process. 
The default recommendation is therefore to set `saved=False` on tables at exit,
rather than resetting the storage. 

4. SIGKILL'ed processes will also leave datasets where `save == False` orphaned.
To clean this up, the new python process can start with:

`>>> Table.load_stored_tables()`

As this will create reference counts and drop them immediately. Saved tables
are unaffected. Should all saved tables be removed - except imports - use:

`>>> [t.save=False for t in Table.load_stored_tables()]`

5. A special and delicate case exists if two terminals use the same import, as
duplicate import is avoided. In the example below P1 and P2 are separate 
processes:

    `P1 >>> t1 = Table.import_file(**config)`
    
    `P1` performed the work to import as set `t1.saved=True`
    
    `P2 >>> t2 = Table.import_file(**config)` 
    
    As `**config` already has been imported by `P1`, `t2` is loaded in `P2`, not imported.

    `P1` has finished its task and sets `saved=False` to exit correctly:
    
    `P1 >>> t1.saved = False`
    `P1 >>> exit() `
    
    Garbage collection of `t1` will now invoke `t1.__del__()` and delete the data.

    `P2 >>> t2.show()`  will return an empty table.

To handle this case `import` and `load` are two separate operations, where 
`import_file` first imports the dataset and then `load`s it.

This means that to get rid of imports as well, the following call must be used:

`>>> [t.save=False for t in Table.load_stored_tables(imports=True)]`

EOM.

**Comment by root-11** [comment]
The changes will be released as 2022.12.0

**Comment by root-11** [comment]
TLDR:

|objective| call |
|---|---|
|Reset the storage| `Table.reset_storage()` |
|Drop temp files after SIGKILL| `Table.load_stored_tables()`|
|Drop all tables except imports | `[t.save=False for t in Table.load_stored_tables()]`|
|Dtop all tables incl. imports | `[t.save=False for t in Table.load_stored_tables(imports=True)]`|

**Comment by root-11** [comment]
Solved since 2023.6
